### PR TITLE
Add trainer and rep to user track

### PR DESCRIPTION
**Pull request by ErikSchierboom**
- Add some ML Training scaffolding
- Add code tagger page (#6445)
- Remove logger
- Clean up Dashboard
- Read from a real sample
- Remove tiny index files
- Clean up DashboardHeader
- Convert code to files
- Improve logic
- Make tabs more dynamic, adjust link type
- Add locking behaviour
- Adjust types, camelize keys
- Add params, statuses
- Fix the tests
- Tidying styling
- Add all tags
- react-query -> tanstack, update yarn lock
- Adjust `all_tags`
- Pass down all_tags
- Pass down things, set default values
- Add missing endpoint, restructure data
- Add training_data_root path
- Fix http method, fix `update_tags`
- Fix wrong linkname
- Add redirect logic
- Update trainer? method
- Verify trainer of tag updating and add tests
- Add serializers for code tags samples
- Remove order param
- Start work on retrieving code samples
- UI adjustments, add Dashboard view to training_Data/dashboard/index
- Allow retrieving samples
- Add link to edit code tags sample
- Reduce indentation
- Convert sample retrieval to filter
- Guard track trainer
- Update tab names / statuses
- Switch to new statuses
- Add training data API request
- Load tracks
- Add non-optimized trainer
- Add trainer and reputation to user_track
- Add command to update user_track reputation
- Update user_track reputation in migration
- Update user_track reputation when new reputation token is created
- Add command to update trainer field of user_track
- Toggle trainer status

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -352,13 +352,8 @@ def automator?(track = nil)
...
     return true if staff?
-    return reputation_for_track(track) >= MIN_REP_TO_TRAIN_ML if track.present?
-
-    # TODO: optimize this
-    User::ReputationPeriod.
-      where(period: :forever, about: :track, user: self).
-      group(:track_id).
-      sum(:reputation).
-      any? { |_, reputation| reputation >= User::MIN_REP_TO_TRAIN_ML }
+    return UserTrack.for(self, track).trainer? if track.present?
```
Comment (ErikSchierboom) [inline app/models/user.rb]: I've used `UserTrack.for` here to use its caching

File: app/models/user.rb [hunk 2]
```diff
@@ -352,13 +352,8 @@ def automator?(track = nil)
...
+
+    user_tracks.where(trainer: true).exists?
```
Comment (ErikSchierboom) [inline app/models/user.rb]: Do we want an index? I didn't do that as the average number of enrolled tracks will probably be relatively low for most people

File: app/models/user_track.rb [hunk 3]
```diff
@@ -25,6 +25,8 @@ class UserTrack < ApplicationRecord
     primary_key: :user_id,
     inverse_of: :user_track
 
+  scope :trainer, -> { where('reputation >= ?', User::MIN_REP_TO_TRAIN_ML) }
```
Comment (ErikSchierboom) [inline app/models/user_track.rb]: I'm open to better names :)

File: app/views/training_data/external/_locked.html.haml [hunk 4]
```diff
@@ -1,12 +1,16 @@
 .border-1.border-borderColor6.shadow-smZ1.rounded-8.py-16.px-32.mb-32{ class: 'w-100 max-w-[782px]' }
   .flex.flex-col.items-stretch
     %h2.reputation.text-20.leading-paragraph.mb-12.font-semibold.flex.items-center
-      %span.mr-12 Your current reputation is
-      = render ViewComponents::Reputation.new(current_user.formatted_reputation, size: :small, flashy: true)
+      - if max_reputation_track
+        %span.mr-12 Your current max reputation in a track is
+        = render ViewComponents::Reputation.new(max_reputation_track.reputation, size: :small, flashy: true)
```
Comment (ErikSchierboom) [inline app/views/training_data/external/_locked.html.haml:6]: One thing left to do here is to display the track for which the user has this reputation. Something for @ihid or @dem4ron

File: app/controllers/training_data/dashboard_controller.rb [hunk 5]
```diff
@@ -1,11 +1,17 @@
...
   def index
     @training_data_dashboard_params = params.permit(:status, :order, :criteria, :page, :track_slug)
     @statuses = %w[needs_tagging needs_checking needs_checking_admin]
   end
 
+  def become_trainer
+    User::BecomeTrainer.(current_user) unless current_user.trainer?
+
+    redirect_to training_data_root_path
+  end
```
Comment (iHiD) [inline app/controllers/training_data/dashboard_controller.rb]: Maybe move this into the `external` controller, so you don't have to have the exception in the guard and you automatically get the `unless` on L10 sorted for you that way too?

File: app/models/user_track.rb (hunk 3 above)
Comment (iHiD) [inline app/models/user_track.rb]: Should this constant move here now? I think maybe it belongs better in here.

File: app/models/user.rb [hunk 6]
```diff
@@ -352,13 +352,16 @@ def automator?(track = nil)
...
-
-    # TODO: optimize this
-    User::ReputationPeriod.
-      where(period: :forever, about: :track, user: self).
-      group(:track_id).
-      sum(:reputation).
-      any? { |_, reputation| reputation >= User::MIN_REP_TO_TRAIN_ML }
+    return super() && eligible_for_trainer?(track) if track.present?
+
+    super()
```
Comment (iHiD) [inline app/models/user.rb]: You could do `return false unless super()` above this to remove the duplicate call from L355 and L357

Comment (iHiD) [reply app/models/user.rb]:
```suggestion
    return false unless super() 
    return eligible_for_trainer?(track) if track.present?

    true
```

File: app/models/user.rb [hunk 7]
```diff
@@ -352,13 +352,16 @@ def automator?(track = nil)
...
+  end
+
+  def eligible_for_trainer?(track = nil)
+    if track
+      user_tracks.trainer.where(track:).exists?
+    else
+      user_tracks.trainer.exists?
```
Comment (iHiD) [inline app/models/user.rb:364]: Maybe:
```suggestion
    query = user_tracks.trainer
    query = query.where(track:) if track
    query.exists?
```

File: app/models/user.rb [hunk 8]
```diff
@@ -352,13 +352,16 @@ def automator?(track = nil)
...
+    end
```
Comment (iHiD) [inline app/models/user.rb]: Maybe:
```suggestion
    uts = user_tracks.trainer
    uts = query.where(track:) if track
    uts.exists?
```

File: app/views/training_data/external/_locked.html.haml [hunk 9]
```diff
@@ -15,5 +19,5 @@
           How does this work?
       .font-medium.text-textColor6.ml-auto
         Ability to use Training Hub unlocks at
```
Comment (iHiD) [inline app/views/training_data/external/_locked.html.haml]:
```suggestion
        Ability to use Training Hub unlocks when you reach
```

File: app/views/training_data/external/_locked.html.haml [hunk 10]
```diff
@@ -15,5 +19,5 @@
...
-        = link_to "#{profile_min_reputation} rep", doc_url(:using, 'product/reputation'), class: 'text-gradient !font-semibold'
-
+        = link_to "#{track_min_reputation} rep", doc_url(:using, 'product/reputation'), class: 'text-gradient !font-semibold'
+        in a track
```
Comment (iHiD) [inline app/views/training_data/external/_locked.html.haml]:
```suggestion
        in a single track
```

File: app/views/training_data/external/index.html.haml [hunk 11]
```diff
@@ -10,10 +10,11 @@
       %p.unsure Bonus: Helping us out earns reputation!
 
       - if user_signed_in?
-        - if current_user.reputation >= User::MIN_REP_TO_TRAIN_ML
-          = link_to "Get Started in our Training Hub →", "#", class: "btn btn-primary btn-l"
+        - if @eligible
+          = form_tag training_data_become_trainer_path, method: :patch, class: "flex flex-row items-center mt-20" do
+            = button_tag "Become a Trainer and Get Started in our Training Hub →", class: "btn btn-primary btn-l"
```
Comment (iHiD) [inline app/views/training_data/external/index.html.haml]:
```suggestion
            = button_tag "Become a Trainer →", class: "btn btn-primary btn-l"
```

File: db/migrate/20231116124259_add_reputation_to_user_track.rb [hunk 12]
```diff
@@ -0,0 +1,11 @@
+class AddReputationToUserTrack < ActiveRecord::Migration[7.0]
+  def change
+    return if Rails.env.production?
+
+    add_column :user_tracks, :reputation, :integer, null: false, default: 0
+
+    UserTrack.find_each do |user_track|
+      UserTrack::UpdateReputation.(user_track)
+    end
```
Comment (iHiD) [inline db/migrate/20231116124259_add_reputation_to_user_track.rb:9]: This will take a while! Definitely do it in tmux.

**Review by iHiD — APPROVED** [review]
Splattering of comments, but looks good! :)

File: app/models/user.rb (hunk 6 above)
Comment (ErikSchierboom) [reply app/models/user.rb]: I had that but I found `return false unless super()` _really_ hard to read.

File: app/models/user.rb (hunk 8 above)
Comment (ErikSchierboom) [reply app/models/user.rb]: I had that first, but refactored to this 🤷

File: app/controllers/training_data/dashboard_controller.rb [hunk 13]
```diff
@@ -1,19 +1,13 @@
...
 
-  def become_trainer
-    User::BecomeTrainer.(current_user) unless current_user.trainer?
-
-    redirect_to training_data_root_path
-  end
-
   def guard_trainer!
-    return true if current_user&.trainer?
+    return true if current_user.trainer?
```
Comment (iHiD) [inline app/controllers/training_data/dashboard_controller.rb:10]: I think this still needs the `&` as logged in users get here by default before being redirected.

File: app/models/user.rb [hunk 14]
```diff
@@ -352,15 +351,16 @@ def automator?(track = nil)
...
 
   def eligible_for_trainer?(track = nil)
-    uts = user_tracks.trainer
-    uts = query.where(track:) if track
-    uts.exists?
+    if track
+      user_tracks.trainer.where(track:).exists?
+    else
+      user_tracks.trainer.exists?
+    end
```
Comment (iHiD) [inline app/models/user.rb]: Or this.

File: app/models/user.rb [hunk 15]
```diff
@@ -352,15 +351,16 @@ def automator?(track = nil)
 
   def trainer?(track = nil)
     return true if staff?
-    return false unless super() 
-    return eligible_for_trainer?(track) if track.present?
+    return super() && eligible_for_trainer?(track) if track.present?
 
-    true
+    super()
```
Comment (iHiD) [inline app/models/user.rb]: I don't think you meant to revert this.

File: app/controllers/training_data/dashboard_controller.rb (hunk 13 above)
Comment (ErikSchierboom) [reply app/controllers/training_data/dashboard_controller.rb:10]: I thought that the default was to require an authenticated user. If I go to that page without being logged-in, I get redirected to the login page.

**Pull request merged**